app/wasm/message_plugin.go

```diff
@@ -0,0 +1,168 @@
+package wasm
+
+import (
+	"encoding/json"
+	"fmt"
+
+	wasmkeeper "github.com/CosmWasm/wasmd/x/wasm/keeper"
+	wasmvmtypes "github.com/CosmWasm/wasmvm/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	bankkeeper "github.com/cosmos/cosmos-sdk/x/bank/keeper"
+
+	wasmbindings "github.com/osmosis-labs/osmosis/v7/app/wasm/bindings"
+	gammkeeper "github.com/osmosis-labs/osmosis/v7/x/gamm/keeper"
+	gammtypes "github.com/osmosis-labs/osmosis/v7/x/gamm/types"
+)
+
+func CustomMessageDecorator(gammKeeper *gammkeeper.Keeper, bank *bankkeeper.BaseKeeper) func(wasmkeeper.Messenger) wasmkeeper.Messenger {
+	return func(old wasmkeeper.Messenger) wasmkeeper.Messenger {
+		return &MintTokenMessenger{
+			wrapped:    old,
+			bank:       bank,
+			gammKeeper: gammKeeper,
+		}
+	}
+}
+
+type MintTokenMessenger struct {
+	wrapped    wasmkeeper.Messenger
+	bank       *bankkeeper.BaseKeeper
+	gammKeeper *gammkeeper.Keeper
+}
+
+var _ wasmkeeper.Messenger = (*MintTokenMessenger)(nil)
+
+func (m *MintTokenMessenger) DispatchMsg(ctx sdk.Context, contractAddr sdk.AccAddress, contractIBCPortID string, msg wasmvmtypes.CosmosMsg) ([]sdk.Event, [][]byte, error) {
+	if msg.Custom != nil {
+		// only handle the happy path where this is really minting / swapping ...
+		// leave everything else for the wrapped version
+		var contractMsg wasmbindings.OsmosisMsg
+		if err := json.Unmarshal(msg.Custom, &contractMsg); err != nil {
+			return nil, nil, sdkerrors.Wrap(err, "osmosis msg")
+		}
+		if contractMsg.MintTokens != nil {
+			return m.mintTokens(ctx, contractAddr, contractMsg.MintTokens)
+		}
+		if contractMsg.Swap != nil {
+			return m.swapTokens(ctx, contractAddr, contractMsg.Swap)
+		}
+	}
+	return m.wrapped.DispatchMsg(ctx, contractAddr, contractIBCPortID, msg)
+}
+
+func (m *MintTokenMessenger) mintTokens(ctx sdk.Context, contractAddr sdk.AccAddress, mint *wasmbindings.MintTokens) ([]sdk.Event, [][]byte, error) {
+	rcpt, err := parseAddress(mint.Recipient)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	denom, err := GetFullDenom(contractAddr.String(), mint.SubDenom)
+	if err != nil {
+		return nil, nil, sdkerrors.Wrap(err, "mint token denom")
+	}
+	coins := []sdk.Coin{sdk.NewCoin(denom, mint.Amount)}
+
+	err = m.bank.MintCoins(ctx, gammtypes.ModuleName, coins)
+	if err != nil {
+		return nil, nil, sdkerrors.Wrap(err, "minting coins from message")
+	}
+	err = m.bank.SendCoinsFromModuleToAccount(ctx, gammtypes.ModuleName, rcpt, coins)
+	if err != nil {
+		return nil, nil, sdkerrors.Wrap(err, "sending newly minted coins from message")
+	}
+
+	return nil, nil, nil
+}
+
+func (m *MintTokenMessenger) swapTokens(ctx sdk.Context, contractAddr sdk.AccAddress, swap *wasmbindings.SwapMsg) ([]sdk.Event, [][]byte, error) {
+	_, err := performSwap(m.gammKeeper, ctx, contractAddr, swap)
+	return nil, nil, sdkerrors.Wrap(err, "gamm estimate price exact amount out")
```

Comment:
This doesn't look right.

Will add unit tests for swap / mint functionality next.

Reply:
Solved in #14.